spec/index.bs

```diff
@@ -251,6 +251,157 @@ dictionary GPUObjectDescriptorBase {
+    and makes are reminder to itself to check back when this usage is over.
+  3. After the GPU operating on [=Queue timeline=] is done using the buffer,
+    the user agent maps it to memory and resolves the promise.
+
+</div>
+
+## Memory ## {#programming-model-memory}
+
+*This section is non-normative.*
+
+Once a {{GPUDevice}} has been obtained during an application initialization routine,
+we can describe the <dfn dfn>WebGPU platform</dfn> as consisting of the following layers:
+  1. User agent implementing the specification.
+  2. Operating system with low-level native API drivers for this device.
+  3. Actual CPU and GPU hardware.
+
+Each layer of the [=WebGPU platform=] may have different memory types
+that the user agent needs to consider when implementing the specification:
+  - The script-owned memory, such as {{ArrayBuffer}} created by the script,
+    is generally not accessible by a GPU driver.
+  - A user agent may have different processes responsible for running
+    the content and communication to the GPU driver.
+    In this case, it uses inter-process shared memory to transfer data.
+  - Dedicated GPUs have their own memory of high bandwidth,
+    while integrated GPUs typically share memory with the system.
+
+Most [=physical resources=] are allocated in the memory of type
+that is efficient for computation or rendering by the GPU.
+When the user needs to provide new data to the GPU,
+the data may first need to cross the process boundary in order to reach
+the user agent part that communicates with the GPU driver.
+Then it may need to be made visible to the driver,
+which sometimes requires a copy into driver-allocated staging memory.
+Finally, it may need to be transferred to the dedicated GPU memory,
+potentially changing the internal layout into one
+that is most efficient for GPUs to operate on.
+
+All of these transitions are done by the WebGPU implementation of the user agent.
+
+Note: This example describes the worst case, while in practice
+the implementation may not need to cross the process boundary,
+or may be able to expose the driver-managed memory directly to
+the user behind an `ArrayBuffer`, thus avoiding any data copies.
+
+## Resource usage ## {#programming-model-resource-usage}
+
+Buffers and textures can be used by the GPU in multiple ways,
+which can be split into two groups:
+
+: <dfn dfn>Read-only usage</dfn>s
+:: Usages like
+    {{GPUBufferUsage/VERTEX|GPUBufferUsage.VERTEX}} or
+    {{GPUTextureUsage/SAMPLED|GPUTextureUsage.SAMPLED}}
+    don't change the contents of a resource.
+
+: <dfn dfn>Mutating usage</dfn>s
+:: Usages {{GPUBufferUsage/STORAGE|GPUBufferUsage.STORAGE}}
+    do change the contents of a resource.
+
+Issue(gpuweb/gpuweb#296): Consider merging all read-only usages.
+
+Textures may consist of separate mipmap levels and array layers,
+which can be used differently at any given time. For the matter of usage validation,
+we'll call them <dfn dfn>subresource</dfn>s.
```

Comment:
```suggestion
we'll call them <dfn dfn>subresources</dfn>.
```

I'm pretty sure this will still link correctly with `[=subresource=]` (due to auto pluralization detection), but obviously double check.